### ISSUE TITLE
git-imerge crash: MergeFrontier partitioned with inappropriate block

### ISSUE THREAD

**Issue by blitz** [mhagger/git-imerge#88]
While doing an git imerge rebase I get the following error:

```
Attempting automerge of 11-29...failure.
Attempting automerge of 1-1...success.
Attempting automerge of 1-16...success.
Attempting automerge of 1-23...success.
Attempting automerge of 1-27...success.
Attempting automerge of 1-29...success.
Attempting automerge of 11-29...failure.
Attempting automerge of 6-29...failure.
Attempting automerge of 4-29...success.
Attempting automerge of 5-29...success.
Attempting automerge of 6-1...success.
Attempting automerge of 6-15...success.
Attempting automerge of 6-22...success.
Attempting automerge of 6-26...success.
Attempting automerge of 6-28...failure.
Attempting automerge of 6-27...failure.
Attempting automerge of 11-26...success.
Autofilling 1-29...success.
Autofilling 2-29...success.
Autofilling 3-29...success.
Autofilling 4-29...success.
Autofilling 5-1...success.
Autofilling 5-2...success.
Autofilling 5-3...success.
Autofilling 5-4...success.
Autofilling 5-5...success.
Autofilling 5-6...success.
Autofilling 5-7...success.
Autofilling 5-8...success.
Autofilling 5-9...success.
Autofilling 5-10...success.
Autofilling 5-11...success.
Autofilling 5-12...unexpected conflict.  Backtracking...
Autofilling 1-29...success.
Autofilling 2-29...success.
Autofilling 3-29...success.
Autofilling 4-1...success.
Autofilling 4-2...success.
Autofilling 4-3...success.
Autofilling 4-4...success.
Autofilling 4-5...success.
Autofilling 4-6...success.
Autofilling 4-7...success.
Autofilling 4-8...success.
Autofilling 4-9...success.
Autofilling 4-10...success.
Autofilling 4-11...success.
Autofilling 4-12...unexpected conflict.  Backtracking...
Autofilling 1-29...success.
Autofilling 2-29...success.
Autofilling 3-1...success.
Autofilling 3-2...success.
Autofilling 3-3...success.
Autofilling 3-4...success.
Autofilling 3-5...success.
Autofilling 3-6...success.
Autofilling 3-7...success.
Autofilling 3-8...success.
Autofilling 3-9...success.
Autofilling 3-10...success.
Autofilling 3-11...success.
Autofilling 3-12...success.
Autofilling 3-13...success.
Autofilling 3-14...success.
Autofilling 3-15...success.
Autofilling 3-16...success.
Autofilling 3-17...success.
Autofilling 3-18...success.
Autofilling 3-19...success.
Autofilling 3-20...success.
Autofilling 3-21...success.
Autofilling 3-22...success.
Autofilling 3-23...success.
Autofilling 3-24...success.
Autofilling 3-25...success.
Autofilling 3-26...success.
Autofilling 3-27...success.
Autofilling 3-28...success.
Autofilling 3-29 (first way)...success.
Autofilling 3-29 (second way)...success.
The two ways of autofilling 3-29 agree.
Recording autofilled block MergeState('single-binary', tip1='upstream/master', tip2='single-binary', goal='rebase')[0:4,0:30].
Traceback (most recent call last):
  File "/home2/julian.stecklina/local/bin/git-imerge", line 3321, in <module>
    main(sys.argv[1:])
  File "/home2/julian.stecklina/local/bin/git-imerge", line 136, in wrapper
    return f(*args, **kw)
  File "/home2/julian.stecklina/local/bin/git-imerge", line 3199, in main
    merge_state.auto_complete_frontier()
  File "/home2/julian.stecklina/local/bin/git-imerge", line 2257, in auto_complete_frontier
    frontier.auto_expand()
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1412, in auto_expand
    if block.auto_expand_frontier():
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1710, in auto_expand_frontier
    return self.auto_outline_frontier()
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1694, in auto_outline_frontier
    return self.auto_outline_frontier(merge_frontier)
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1694, in auto_outline_frontier
    return self.auto_outline_frontier(merge_frontier)
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1696, in auto_outline_frontier
    f1, f2 = merge_frontier.partition(best_block)
  File "/home2/julian.stecklina/local/bin/git-imerge", line 1342, in partition
    'MergeFrontier partitioned with inappropriate block'
ValueError: MergeFrontier partitioned with inappropriate block
```

**Comment by mhagger** [comment]
Thanks for the bug report.

Did this happen in a public repository that I could clone to try to reproduce the problem? If so, please let us know exactly what you had to run to trigger the error.

**Comment by blitz** [comment]
Unfortunately, this is a private repository. I am happy to use a version that spits out more debug information though.

**Comment by mhagger** [comment]
@blitz, would you please check whether #91 fixes the problem that you experienced? Thanks!

**Comment by blitz** [comment]
Can't reproduce it any more, but unfortunately I don't have the exact repo around. Will copy it to a safe place next time. Sorry.

**Comment by davidchisnall** [comment]
I've encountered a very similar problem (not sure if it's the same thing or a new issue):

```
...
Autofilling 100-614...success.
Autofilling 100-615...success.
Autofilling 100-616...success.
Autofilling 100-617...success.
Autofilling 100-618 (first way)...success.
Autofilling 100-618 (second way)...success.
The two ways of autofilling 100-618 agree.
Recording autofilled block MergeState('upstream', tip1='master', tip2='upstream', goal='merge')[94:101,209:619].
Attempting automerge of 65-5641...failure.
Attempting automerge of 58-3294...failure.
Attempting automerge of 76-3293...failure.
Attempting automerge of 66-2821...failure.
Attempting automerge of 94-2820...success.
Autofilling 77-2820...unexpected conflict.  Backtracking...
Autofilling 77-2819...unexpected conflict.  Backtracking...
Autofilling 77-2818...unexpected conflict.  Backtracking...
Autofilling 77-2817...unexpected conflict.  Backtracking...
Autofilling 77-2816...unexpected conflict.  Backtracking...
Autofilling 77-2815...unexpected conflict.  Backtracking...
Autofilling 77-2814...unexpected conflict.  Backtracking...
Autofilling 77-2813...unexpected conflict.  Backtracking...
Autofilling 77-2812...unexpected conflict.  Backtracking...
Autofilling 77-2811...unexpected conflict.  Backtracking...
Autofilling 77-2810...unexpected conflict.  Backtracking...
Autofilling 77-2809...unexpected conflict.  Backtracking...
Autofilling 77-2808...unexpected conflict.  Backtracking...
...
Autofilling 77-1841...unexpected conflict.  Backtracking...
Autofilling 77-1840...Traceback (most recent call last):
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 3947, in <module>
    main(sys.argv[1:])
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 3928, in main
    cmd_continue(parser, options)
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 3499, in cmd_continue
    merge_state.auto_complete_frontier()
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 2459, in auto_complete_frontier
    frontier.auto_expand()
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 1603, in auto_expand
    if block.auto_expand_frontier():
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 1901, in auto_expand_frontier
    return self.auto_outline_frontier()
  File "/Users/theraven/tmp/git-imerge/git-imerge", line 1885, in auto_outline_frontier
    return self.auto_outline_frontier(merge_frontier)
```
The last line then repeats a lot of times.  I've spent about a week and a half of CPU time on this merge so far, so if there's a way to recover without starting again then that would be very helpful.

I'm not sure if it's related, but I ran a `git gc` last night because the individual merges were taking around 20 seconds (now they're down to 1-2 seconds) and at that speed it would have taken about six months to finish the merge.

The upstream repo is:

https://github.com/llvm-mirror/clang

And I'm merging into:

https://github.com/ctSRD-CHERI/clang/

Note: I'm having to use the loosen-first-parent-rule branch of git-imerge, because master fails to find a common parent.

**Comment by davidchisnall** [comment]
Actually, my issue might be caused by running out of stack space in the recursive call to `auto_outline_frontier`: I don't know how big the Python stack can be, but 1000 stack frames might be a few too many.

**Comment by davidchisnall** [comment]
It appears that my issue was, indeed, caused by the fact that Python doesn't do tail call optimisation.  The following ugly hack fixed it for me:

```diff
--- a/git-imerge
+++ b/git-imerge
@@ -1882,7 +1882,7 @@ class Block(object):
             # One of the merges that we expected to succeed in
             # fact failed.
             merge_frontier.remove_failure(e.i1, e.i2)
-            return self.auto_outline_frontier(merge_frontier)
+            raise Exception(merge_frontier)
         else:
             f1, f2 = merge_frontier.partition(best_block)
             if f1:
@@ -1898,7 +1898,17 @@ class Block(object):
         elif merge_state.goal == 'full':
             return self.auto_fill_micromerge()
         else:
-            return self.auto_outline_frontier()
+            try:
+                return self.auto_outline_frontier()
+            except Exception as mf:
+                merge_frontier = mf.args[0]
+                while True:
+                    try:
+                        return self.auto_outline_frontier(merge_frontier)
+                    except Exception as mf:
+                        merge_frontier = mf.args[0]
+
+
 
     # The codes in the 2D array returned from create_diagram()
     MERGE_UNKNOWN = 0
```

A similar but less ugly fix is probably required, as backtracking is currently limited to a depth of around 1,000 by the Python default stack size (and can't easily grow larger without running out of C stack space).

**Comment by mhagger** [comment]
> The last line then repeats a lot of times. I've spent about a week and a half of CPU time on this merge so far [...]
> 
> I'm not sure if it's related, but I ran a `git gc` last night because the individual merges were taking around 20 seconds (now they're down to 1-2 seconds) and at that speed it would have taken about six months to finish the merge.

That's amazing (and scary!) I'm curious to what extent your merge is so slow simply because it is gigantic, vs. whether you are hitting a pathological case. I suspect the latter.

These `UnexpectedMergeFailure` exceptions are meant to be just that, unexpected. They occur when it is possible to merge a whole series of commits together, but then when it tries to merge the commits one by one, there is a failure somewhere along the way. In other words, your merge is violating [this approximate assumption](https://github.com/mhagger/git-imerge/blob/master/git-imerge#L985-L987).

The backtracking that you see is `git-imerge`'s way of recovering. Because this is expected to be an unusual occurrence, this code path is not written in a way that would be appropriate if it were expected that the assumption would be violated not a few times, but more than ~1000 times.

The repeated backtracking that you see, I believe, is the very first step of outlining a block of commits that are all expected to be doable. This is the merge of one commit from one side of the block with all of the commits on the other side of the block. When that merge unexpectedly fails, it takes only one step backwards and tries again. It is quite possible that this is making the overall algorithm regress to O(N²) or something.

Your "ugly" fix is pretty much what my first impulse would have been (though probably with a custom exception type built for this purpose). Feel free to write it up as a PR. But additionally, I hypothesize that an algorithmic improvement could reduce your pain dramatically. For example, `MergeFrontier.remove_failure` could do a bisection step to more quickly find a less ambitious merge that succeeds.

I'm really curious to hear whether your merge was/will be successful eventually.

**Comment by davidchisnall** [comment]
The merge did succeed last night.  It looks as if I'd missed some commit markers in one of the previous merges and so the next one ended up being nonsense and git-imerge gave up because of this.  Backtracking around 2000 upstream revisions undid this error and let me fix it properly (well, in a way that didn't conflict - I need to wait for the LLVM merge to finish before I can tell whether it's actually working).  The backtracking took about 45 minutes, which was not that bad in comparison to a merge that took two and a half weeks overall.  Almost all of this time is spent autofilling, and so anything that could reduce the number of those that are done would help a lot.  It would also possibly help if git-imerge could time the pairwise merges and run git gc periodically if the time has increased a lot.  I've found that it creeps from about 2 seconds to about 30 seconds in the larger merge and then drops down a lot after a gc.  The gc isn't deleting much, but the compression stage appears to dramatically reduce the amount of I/O the git processes do (and means that they're sitting at 100% CPU usage, rather than mostly blocked on I/O, even with a fast SSD).
